torch/_inductor/runtime/static_cuda_launcher.py

```diff
@@ -0,0 +1,217 @@
+        converts it into a single char encoding. _StaticCudaLauncher
+        will switch on this char to figure out what type the underlying
+        value should be passed to the triton kernel as.
+        """
+        if ty[0] == "*":
+            return "O"
+        elif ty == "nvTmaDesc":
+            raise NotImplementedError("nvTmaDesc kernels are not yet supported")
+        return {
+            "i1": "i",
+            "i8": "b",
+            "i16": "h",
+            "i32": "i",
+            "i64": "l",
+            "u1": "I",
+            "u8": "B",
+            "u16": "H",
+            "u32": "I",
+            "u64": "K",
+            "fp16": "f",
+            "bf16": "f",
+            "fp32": "f",
+            "f32": "f",
+            "fp64": "d",
+            # TODO handle nvTmaDesc/CUtensormap
+        }[ty]
+
+    def arg_ty_from_signature(self, src: ASTSource) -> tuple[str, OrderedSet[int]]:
+        def index_key(i: Any) -> int:
+            return src.fn.arg_names.index(i) if isinstance(i, str) else i
+
+        signature = {index_key(key): value for key, value in src.signature.items()}
+        constants = [index_key(key) for key in getattr(src, "constants", dict())]
+        # Despite requiring them to be passed in, the triton CUDA launcher
+        # completely ignores the constexprs passed into it when generating code.
+        # So we can ignore them here too
+        params = []
+
+        constant_idxs: OrderedSet[int] = OrderedSet()
+        for i in sorted(signature.keys()):
+            ty = signature[i]
+            # In newer triton versions, constants are passed in to signature with type `constexpr`
+            # In older triton versions, there can be constants in src.constants that are not `constexpr` in signature
+            # so we check both here
+            if ty == "constexpr" or i in constants:
+                constant_idxs.add(i)
+            else:
+                params.append(self.extract_type(ty))
+        return "".join(params), constant_idxs
+
+    def run(
+        self, grid: tuple[int, ...], stream: int, *args: Unpack[tuple[object, ...]]
+    ) -> None:
+        """Actually run the kernel at runtime. This function is the hot codepath."""
+        from torch._C import _StaticCudaLauncher
+
+        # Assert load_kernel() has been called and args match
+        assert self.function is not None
+
+        # TODO: actually, if the args *don't* match, we probably should
+        # throw an exception. But if inductor is the only one calling this
+        # thing, it should always match.
+        # Get rid of constants before passing to cubin launcher
+
+        # TODO: is this (and the check below) slow to do at runtime? The thing is,
+        # we already spend the time in CachingAutotuner.launch() to massage the arguments
+        # properly anyways so this isn't exactly slower than that...
+        args = tuple(args[i] for i in range(len(args)) if i not in self.constant_idxs)
```

Comment:
Question for folks here. I don't like the idea of manipulating `args` at runtime like this, and I suppose that's what the `exec` in TritonCompileResult aims to handle, by codegenning python and execcing a perfect function to pass. However, in CachingAutotuner.run, we manipulate the args by adding `constexprs` and stuff to the list of runtime args anyways (see `_get_args_with_constexprs(args, launcher)`)

.... so it seems like it's not actually particularly slow to manipulate the args at kernel runtime. So maybe it's not that necessary to optimize? Or am I not reading this correctly?